mesonbuild/interpreter/primitives/string.py

```diff
@@ -132,7 +132,19 @@ def substring_method(self, args: T.Tuple[T.Optional[int], T.Optional[int]], kwar
     @InterpreterObject.method('to_int')
     def to_int_method(self, args: T.List[TYPE_var], kwargs: TYPE_kwargs) -> int:
         try:
-            return int(self.held_object)
+            s = self.held_object.strip()
+            s_unsigned = s.lstrip('+-').lower()
+
+            if s_unsigned.startswith('0x'):
+                base = 16
+            elif s_unsigned.startswith('0o'):
+                base = 8
+            elif s_unsigned.startswith('0b'):
+                base = 2
+            else:
+                base = 10
+
+            return int(s, base)
```

Comment:
```suggestion
            return int(s, base=0)
```

Reply:
This cannot be done unilaterally because of backwards compatibility. Before this, meson treats "010" as being "10" (decimal), and Python 3's int() with no base (or base=10) also treats it as decimal. However, Python 3's int() with base=0 raises ValueError for strings with leading zeros because it disallows leading zeros for decimal numbers when auto-detecting the base. So while we can indeed use base=0 to simplify some cases, we still need an explicit check for the leading-zero case to maintain backward compatibility.